ethcore/src/engines/mod.rs

```diff
@@ -52,15 +52,34 @@ use spec::CommonParams;
 use transaction::{self, UnverifiedTransaction, SignedTransaction};
 
 use ethkey::Signature;
-use parity_machine::{Machine, LocalizedMachine as Localized};
+use parity_machine::{Machine, LocalizedMachine as Localized, TotalScoredHeader};
 use ethereum_types::{H256, U256, Address};
 use unexpected::{Mismatch, OutOfBounds};
 use bytes::Bytes;
+use types::ancestry_action::AncestryAction;
 
 /// Default EIP-210 contract code.
 /// As defined in https://github.com/ethereum/EIPs/pull/210
 pub const DEFAULT_BLOCKHASH_CONTRACT: &'static str = "73fffffffffffffffffffffffffffffffffffffffe33141561006a5760014303600035610100820755610100810715156100455760003561010061010083050761010001555b6201000081071515610064576000356101006201000083050761020001555b5061013e565b4360003512151561008457600060405260206040f361013d565b61010060003543031315156100a857610100600035075460605260206060f361013c565b6101006000350715156100c55762010000600035430313156100c8565b60005b156100ea576101006101006000350507610100015460805260206080f361013b565b620100006000350715156101095763010000006000354303131561010c565b60005b1561012f57610100620100006000350507610200015460a052602060a0f361013a565b600060c052602060c0f35b5b5b5b5b";
 
+/// Fork choice.
+#[derive(Debug, PartialEq, Eq)]
+pub enum ForkChoice {
+	/// Choose the new block.
+	New,
+	/// Choose the current best block.
+	Old,
+}
+
+/// Finalization related information.
+#[derive(Debug, PartialEq, Eq)]
+pub struct FinalizationInfo {
+	/// Whether the route from current best block to common ancestor is finalized.
+	pub is_old_route_finalized: bool,
+	/// Whether the route from new block's parent to common ancestor is finalized.
```

Comment:
do these comments refer to the _whole_ route or not? These properties don't seem particularly useful to me.

Reply:
So with the following route:

```
   B - C
  /
A
  \
   D - E - (F)
```

If `C` is the current best block, and we insert block `F` then:

* `is_old_route_finalized` check finalization in `B - C`.
* `is_new_parent_route_finalized` check finalization in `D - E`, without `F`.

We only use `is_old_route_finalized` for most cases (like in Casper), and I agree `is_new_parent_route_finalized` is probably not that useful. But I just included it because we have this information, and it nearly hurts nothing to expose that.

I'm both okay whether we include it or not.

Reply:
`is_old_route_finalized` checks how many blocks back?
`is_new_parent_route_finalized` is mutually exclusive with `is_old_route_finalized`. This should be an enum.

Reply:
It only checks blocks in the reorg route. So as in the example https://github.com/paritytech/parity/pull/8401#discussion_r185317727, `is_old_route_finalized` only checks `B - C` -- blocks from best to common ancestor. `is_new_parent_route_finalized` checks `D - E` -- inserting block to common ancestor. They're not mutually exclusive.